osd: fix shard-threads cannot wakeup bug

jianwei1216 opened the pull request:
Reproduce:
(1) ceph cluster not running any client IO
(2) only ceph osd in osd.14 operation

Reason:
(1) one shard-queue has three shard-threads
(2) one or some PeeringOp's epoch > osdmap's epoch held by current osd,
    and these PeeringOp _add_slot_waiter()
(3) shard-queue become empty and three shard-threads cond.wait()
(4) new osdmap consume and it _wake_pg_slot()
    Problem in here
	1> OSDShard::consume() exec loop all pg's slot wait
 	   and requeue more than one PeeringOp to shard-queue
        2> but it only notify one shard-thread to wakeup,
           the other two shard-threads continue cond.wait()
	3> OSD::ShardedOpWQ::_enqueue() found the shard-queue not empty
	   and not notify all shard-thread to wakeup

     In a period of time, only one shard-thread of 3 shard-threads is running.

Signed-off-by: Jianwei Zhang <jianwei1216@qq.com>
Change-Id: I4617db2fd95082007e6d9fa2b60f17f2a6296b5b


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

jianwei1216 commented:
(1) one shard-queue has three shard-threads
```
# grep 'osd.14 op_wq(1) _process' ceph-osd.14.log.2 |awk '{print $2}'|sort -u
7fc86e8d4700   ### t1
7fc8710d9700   ### t2
7fc8738de700   ### t3

# grep -w 7fc86e8d4700 op_wq.1.all |wc -l
104533   ### t1   print 104k log items
[root@lenovo15 ceph]# grep -w 7fc8710d9700 op_wq.1.all |wc -l
4792       ### t2   only print 4k log items
[root@lenovo15 ceph]# grep -w 7fc8738de700 op_wq.1.all |wc -l
1109       ### t3   only print 1k log items
```

The amount of tasks performed by the 3 shard threads is completely unequal

jianwei1216 commented:
(2) one or some PeeringOp's epoch > osdmap's epoch held by current osd,
and these PeeringOp _add_slot_waiter()

[https://github.com/ceph/ceph/blob/octopus/src/osd/OSD.cc#L10668](url)

jianwei1216 commented:
(2) one or some PeeringOp's epoch > osdmap's epoch held by current osd,and these PeeringOp _add_slot_waiter()

> void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)

![image](https://user-images.githubusercontent.com/19184136/135411539-625a9161-bc99-4015-abe9-1c35cc6af0cc.png)

jianwei1216 commented:
(2) one or some PeeringOp's epoch > osdmap's epoch held by current osd,and these PeeringOp _add_slot_waiter()

> void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)

![image](https://user-images.githubusercontent.com/19184136/135411539-625a9161-bc99-4015-abe9-1c35cc6af0cc.png)

jianwei1216 commented:
(3) shard-queue become empty and three shard-threads cond.wait()

![image](https://user-images.githubusercontent.com/19184136/135412522-a92da754-9abe-4fcd-96f1-4b436cb27bba.png)



```
2307 2021-09-30T10:04:43.232+0800 7f2421dbe700 20 osd.14 op_wq(1) _process empty q, waiting
2308 2021-09-30T10:04:47.130+0800 7f2421dbe700  0 osd.14 op_wq(1) _process sdata->sdata_cond.wait(wait_lock) latency is 3899 ms

2424 2021-09-30T10:04:43.232+0800 7f24245c3700 20 osd.14 op_wq(1) _process empty q, waiting
2425 2021-09-30T10:04:47.130+0800 7f24245c3700  0 osd.14 op_wq(1) _process sdata->sdata_cond.wait(wait_lock) latency is 3898 ms
```

jianwei1216 commented:
(4) new osdmap consume and it _wake_pg_slot()

> void OSDShard::consume_map

![image](https://user-images.githubusercontent.com/19184136/135412871-c08dad0d-560b-4499-a414-fa6bf229032f.png)

loop pg slot maybe requeue more than one pg's PeeringOp

> void OSDShard::_wake_pg_slot

![image](https://user-images.githubusercontent.com/19184136/135413133-cb57409c-9436-4b88-95c9-fa2bb4f453f1.png)

loop  slot->to_process, slot->wait, slot->want_peering  queues maybe requeue more than one Op

> void OSDShard::consume_map
![image](https://user-images.githubusercontent.com/19184136/135413007-5657505e-0f31-49be-83ed-8b40d531d52e.png)

but it only notify one shard-thread to wakeup

jianwei1216 commented:
3> OSD::ShardedOpWQ::_enqueue() found the shard-queue not empty and not notify all shard-thread to wakeup

![image](https://user-images.githubusercontent.com/19184136/135413933-91ea14b2-ffe9-4d44-9215-a420b34c776c.png)

the shard-queue is not empty, so it does not wakeup all shard-threads

jianwei1216 commented:
> (1) one shard-queue has three shard-threads
> 
> ```
> # grep 'osd.14 op_wq(1) _process' ceph-osd.14.log.2 |awk '{print $2}'|sort -u
> 7fc86e8d4700   ### t1
> 7fc8710d9700   ### t2
> 7fc8738de700   ### t3
> 
> # grep -w 7fc86e8d4700 op_wq.1.all |wc -l
> 104533   ### t1   print 104k log items
> [root@lenovo15 ceph]# grep -w 7fc8710d9700 op_wq.1.all |wc -l
> 4792       ### t2   only print 4k log items
> [root@lenovo15 ceph]# grep -w 7fc8738de700 op_wq.1.all |wc -l
> 1109       ### t3   only print 1k log items
> ```
> 
> The amount of tasks performed by the 3 shard threads is completely unequal

test after add this patch
![image](https://user-images.githubusercontent.com/19184136/135414329-4f61dcf8-81eb-4a45-99d7-45a87d8983ca.png)

The amount of tasks performed by the 3 shard threads is almost the same

jianwei1216 commented:
add @neha-ojha @tchaikov

jianwei1216 commented:
https://tracker.ceph.com/issues/52781?next_issue_id=52780

jianwei1216 commented:
![image](https://user-images.githubusercontent.com/19184136/135420911-afc671c5-feba-4e80-9c6d-fb5e1682aaf5.png)

① shard-thread-3  _add_slot_waiter 2.2c2 , then shard-thread-3 found shard-queue-1 empty and cond.wait()
② shard-thread-2  found shard-queue-1 empty and cond.wait() at the same time
③ shard-thread-1  found shard-queue-1 empty and cond.wait() at the same time
     shard-thread-1 17:32:06.674 wakeup and handle the 2.2c2  (it is wakeup by OSDShard::consume_map)

jianwei1216 commented:
![image](https://user-images.githubusercontent.com/19184136/135422153-21ff167b-ff46-40ce-8533-6308376e8b85.png)

shard-thread-1 17:32:06.674 wakeup by OSDShard::consume_map

neha-ojha commented:
Hi @jianwei1216, I noticed that you have opened this PR against octopus, is https://tracker.ceph.com/issues/52781 only seen in octopus? If the fix is relevant to all releases, let's open a PR against master and we'll backport the fix to octopus.

tchaikov commented:
@jianwei1216 thanks for your fix. i think it also applies to master, so i

1. created a tracker ticket to facilitate the backport
2. retargeted it to master
3. updated the commit message to link the commit to the tracker ticket
4. updated the title of the commit to summarize the change as "fix shard-threads cannot wakeup bug" is not specific enough, IMHO.

hopefully it's fine by you.

jianwei1216 commented:
> @jianwei1216 thanks for your fix. i think it also applies to master, so i
> 
> 1. retargeted it to master
> 2. updated the commit message to link the commit to the tracker ticket
> 3. updated the title of the commit to summarize the change as "fix shard-threads cannot wakeup bug" is not specific enough, IMHO.
> 
> hopefully it's fine by you.

Thank you.
took a holiday a few days ago.

No problem.
